packages/grafbase-sdk/tests/unit/auth.openid.test.ts

```diff
@@ -0,0 +1,465 @@
+        }
+      }
+    })
+
+    expect(cfg.toString()).toMatchInlineSnapshot(`
+      "extend schema
+        @auth(
+          providers: [
+            { type: oidc, issuer: "{{ env.ISSUER_URL }}" }
+          ]
+          rules: [
+            { allow: private, operations: [get, list, read] }
+          ]
+        )"
+    `)
+  })
+
+  it('renders a provider with owner access', () => {
+    const clerk = auth.OpenIDConnect({
+      issuer: '{{ env.ISSUER_URL }}'
+    })
+
+    const cfg = config({
+      schema: g,
+      auth: {
+        providers: [clerk],
+        rules: (rules) => {
+          rules.owner()
+        }
+      }
+    })
+
+    expect(cfg.toString()).toMatchInlineSnapshot(`
+      "extend schema
+        @auth(
+          providers: [
+            { type: oidc, issuer: "{{ env.ISSUER_URL }}" }
+          ]
+          rules: [
+            { allow: owner }
+          ]
+        )"
+    `)
+  })
+
+  it('renders a provider with owner access and custom operations', () => {
+    const clerk = auth.OpenIDConnect({
+      issuer: '{{ env.ISSUER_URL }}'
+    })
+
+    const cfg = config({
+      schema: g,
+      auth: {
+        providers: [clerk],
+        rules: (rules) => {
+          rules.owner().create()
+        }
+      }
+    })
+
+    expect(cfg.toString()).toMatchInlineSnapshot(`
+      "extend schema
+        @auth(
+          providers: [
+            { type: oidc, issuer: "{{ env.ISSUER_URL }}" }
+          ]
+          rules: [
+            { allow: owner, operations: [create] }
+          ]
+        )"
+    `)
+  })
+
+  it('renders a provider with groups access', () => {
+    const clerk = auth.OpenIDConnect({
+      issuer: '{{ env.ISSUER_URL }}'
+    })
+
+    const cfg = config({
+      schema: g,
+      auth: {
+        providers: [clerk],
+        rules: (rules) => {
+          rules.groups(['backend', 'admin'])
+        }
+      }
+    })
+
+    expect(cfg.toString()).toMatchInlineSnapshot(`
+      "extend schema
+        @auth(
+          providers: [
+            { type: oidc, issuer: "{{ env.ISSUER_URL }}" }
+          ]
+          rules: [
+            { allow: groups, groups: ["backend", "admin"] }
+          ]
+        )"
+    `)
+  })
+
+  it('renders a provider with groups access and custom operations', () => {
+    const clerk = auth.OpenIDConnect({
+      issuer: '{{ env.ISSUER_URL }}'
+    })
+
+    const cfg = config({
+      schema: g,
+      auth: {
+        providers: [clerk],
+        rules: (rules) => {
+          rules.groups(['backend', 'admin']).delete()
+        }
+      }
+    })
+
+    expect(cfg.toString()).toMatchInlineSnapshot(`
+      "extend schema
+        @auth(
+          providers: [
+            { type: oidc, issuer: "{{ env.ISSUER_URL }}" }
+          ]
+          rules: [
+            { allow: groups, groups: ["backend", "admin"], operations: [delete] }
+          ]
+        )"
+    `)
+  })
+
+  it('renders multiple providers like a champ', () => {
```

Comment:
Although the schema is valid, we currently only support one provider, so this would cause parsing error. Would it be possible to have a constraint on the size of list == 1 until we implement multiple providers?

Reply:
We can do that. I asked @fbjork about it before doing so, and he said it's possible to add multiple providers in the SDL so therefore we also do it in the typescript config. I get it that it breaks things under the surface, and it's easy enough to prevent this, so I'll do it.

Reply:
Fixed now to be an array of size one: https://github.com/grafbase/grafbase/pull/392/commits/ef0ab6b1d373b48702cd1fcc79a1b2c959879eb9